### PR TITLE
Implement JSONRPC server

### PR DESCRIPTION
**Pull request by omerfirmak**
Implements the server and a HTTP transport layer

extracted from #511 to make it easier to review

### CODE REVIEW

File: jsonrpc/server_test.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+		"no error return": {
+			handler: func(param1, param2 int) (any, int) { return 0, 0 },
+			want:    "second return value must be a *jsonrpc.Error",
+		},
+	}
+
+	for desc, test := range tests {
+		err := server.RegisterMethod("method", test.paramNames, test.handler)
+		assert.EqualError(t, err, test.want, desc)
+	}
```
Comment (IronGauntlets) [inline jsonrpc/server_test.go:49]: Please create sub-tests with `desc` as their name.

File: jsonrpc/server_test.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+	err := server.RegisterMethod("method", nil, func(param1, param2 int) (int, *jsonrpc.Error) { return 0, nil })
+	assert.NoError(t, err)
```
Comment (IronGauntlets) [inline jsonrpc/server_test.go:52]: Please name this test case and add it as a sub test.

File: jsonrpc/server_test.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+			}{*num * 2}, nil
+		}))
+
+	tests := []struct {
+		req string
+		res string
+	}{
+		{
+			req: `{]`,
+			res: `{"jsonrpc":"2.0","error":{"code":-32600,"message":"invalid character ']' looking for beginning of object key string"},"id":null}`,
```
Comment (IronGauntlets) [inline jsonrpc/server_test.go:73]: This test case should return `-32700` which is for invalid json. `-32600` is for well-formed json but an invalid request.

File: jsonrpc/server.go [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+}
+
+// RegisterMethod verifies and creates and endpoint that server recognizes.
+//
+// - name is the method name
+// - handler is the function to be called when a request is received for the
+// associated method. It should have (any, error) as it's return type
+// - paramNames are the names of parameters in the order that they are expected
+// by the handler
+func (s *Server) RegisterMethod(name string, params []Parameter, handler any) error {
```
Comment (IronGauntlets) [inline jsonrpc/server.go:111]: According to the comment you should declare the following type:

```go
type HandlerFn func(args ...any) (any, error)
```

You can unexport it if needed. Then change the signature of the function as follows:
```suggestion
func (s *Server) RegisterMethod(name string, params []Parameter, handler HandlerFn) error {
```

This way you can reduce the use of reflect package here.

File: jsonrpc/server.go [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+	handlerT := reflect.TypeOf(handler)
+	if handlerT.Kind() != reflect.Func {
+		return errors.New("handler must be a function")
+	}
+	if len(params) > 0 && handlerT.NumIn() != len(params) {
+		return errors.New("number of function params and param names must match")
+	}
```
Comment (IronGauntlets) [inline jsonrpc/server.go:118]: parameters are optional so why are we ensuring `len(params) > 0`?

File: jsonrpc/server_test.go [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+			res: `{"jsonrpc":"2.0","error":{"code":44,"message":""},"id":22}`,
+		},
+		{
+			req: `{"jsonrpc" : "2.0", "method" : "method",
+					"params" : { "shouldError" : true } , "id" : 22}`,
+			res: `{"jsonrpc":"2.0","error":{"code":-32602,"message":"missing non-optional param"},"id":22}`,
+		},
+		{
+			req: `[]`,
+			res: `[]`,
```
Comment (IronGauntlets) [inline jsonrpc/server_test.go:117]: This is an incorrect test case. The specification says the following in the last paragraph of section 6: 
```
If there are no Response objects contained within the Response array as it is to be sent
to the client, the server MUST NOT return an empty Array and should return nothing at all.
``` 

Nothing should be returned to the client.

File: jsonrpc/server_test.go [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+	server := jsonrpc.NewServer()
+	require.NoError(t, server.RegisterMethod("method", []jsonrpc.Parameter{{Name: "num"}, {Name: "shouldError", Optional: true}, {Name: "msg", Optional: true}},
+		func(num *int, shouldError bool, msg string) (any, *jsonrpc.Error) {
+			if shouldError {
+				return nil, &jsonrpc.Error{Code: 44, Message: msg}
+			}
+			return struct {
+				Doubled int `json:"doubled"`
+			}{*num * 2}, nil
+		}))
```
Comment (IronGauntlets) [inline jsonrpc/server_test.go:65]: Please separate this into a sub test case with a name

File: jsonrpc/server_test.go [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+					"params" : { "num" : 5 } , "id" : 5},
+					{"jsonrpc" : "2.0", "method" : "fail",
+					"params" : { "num" : 5 } , "id" : 7},
+					{"jsonrpc" : "2.0", "method" : "method",
+					"params" : { "num" : 44 } , "id" : 6}]`,
+			res: `[{"jsonrpc":"2.0","result":{"doubled":10},"id":5},{"jsonrpc":"2.0","error":{"code":-32601,"message":"method not found"},"id":7},{"jsonrpc":"2.0","result":{"doubled":88},"id":6}]`,
+		},
+	}
+
+	for _, test := range tests {
```
Comment (IronGauntlets) [inline jsonrpc/server_test.go:142]: Please separate this into sub test with names.

File: jsonrpc/server_test.go [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+		res, err := server.Handle([]byte(test.req))
+		assert.NoError(t, err)
```
Comment (IronGauntlets) [inline jsonrpc/server_test.go:144]:
```suggestion
		require.NoError(t, err)
```

Otherwise, there will be a panic on the next line is `res` is nil.

**Review by IronGauntlets — CHANGES_REQUESTED** [review]

File: jsonrpc/server.go (hunk 5 above)
Comment (omerfirmak) [reply jsonrpc/server.go:118]: We are not ensuring that. We are simply checking that the handler accepts that many arguments; only if user provides params while registering

File: jsonrpc/server_test.go (hunk 3 above)
Comment (omerfirmak) [reply jsonrpc/server_test.go:73]: yeap, fixed

File: jsonrpc/server.go (hunk 4 above)
Comment (omerfirmak) [reply jsonrpc/server.go:111]: I tried to do something similar but sadly it does not work;

``` golang
type HandlerFn func(args ...any) (any, *Error)


node/node.go:91:30: cannot use rpcHandler.ChainId (value of type func() (*felt.Felt, *jsonrpc.Error)) as jsonrpc.HandlerFn value in struct literal
node/node.go:92:34: cannot use rpcHandler.BlockNumber (value of type func() (uint64, *jsonrpc.Error)) as jsonrpc.HandlerFn value in struct literal
node/node.go:93:41: cannot use rpcHandler.BlockNumberAndHash (value of type func() (*rpc.BlockNumberAndHash, *jsonrpc.Error)) as jsonrpc.HandlerFn value in struct literal
node/node.go:94:79: cannot use rpcHandler.GetBlockWithTxHashes (value of type func(id *rpc.BlockId) (*rpc.BlockWithTxHashes, *jsonrpc.Error)) as jsonrpc.HandlerFn value in struct literal
node/node.go:95:74: cannot use rpcHandler.GetBlockWithTxs (value of type func(id *rpc.BlockId) (*rpc.BlockWithTxs, *jsonrpc.Error)) as jsonrpc.HandlerFn value in struct literal
node/node.go:96:87: cannot use rpcHandler.GetTransactionByHash (value of type func(hash *felt.Felt) (*rpc.Transaction, *jsonrpc.Error)) as jsonrpc.HandlerFn value in struct literal
```

comment is outdated tho, updating

File: jsonrpc/server_test.go (hunk 7 above)
Comment (omerfirmak) [reply jsonrpc/server_test.go:65]: that is not a testcase, hence the `require`. Added descriptions to following table tests.

File: jsonrpc/server_test.go (hunk 6 above)
Comment (omerfirmak) [reply jsonrpc/server_test.go:117]: Thanks, fixed

**Comment by omerfirmak** [comment]
Changes according to comments https://github.com/NethermindEth/juno/compare/ac2c4b8f7325de759e93f4accdbb8fbc8aeb4ae7..32946af23707bff666eaeaddc480f97a2098e522

File: jsonrpc/server.go [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+		return nil, &Error{Code: InvalidJson, Message: err.Error()}
+	} else if err = req.isSane(); err != nil {
+		return nil, &Error{Code: InvalidRequest, Message: err.Error()}
+	}
+	return req, nil
+}
+
+func (r *request) isSane() error {
+	if r.Version != "2.0" {
+		return errors.New(fmt.Sprintf("unsupported RPC request version"))
```
Comment (jelilat) [inline jsonrpc/server.go:59]:
```suggestion
		return errors.New("unsupported RPC request version")
```

Comment (jelilat) [reply jsonrpc/server.go:59]: Or probably this 

```suggestion
		return fmt.Errorf("unsupported RPC request version: %s", r.Version)
```

File: jsonrpc/server.go [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+	if handlerT.Kind() != reflect.Func {
+		return errors.New("handler must be a function")
+	}
+	if len(params) > 0 && handlerT.NumIn() != len(params) {
+		return errors.New("number of function params and param names must match")
+	}
+	if handlerT.NumOut() != 2 {
+		return errors.New("handler must return 2 values")
+	}
+	if handlerT.Out(1) != reflect.TypeOf((*Error)(nil)) {
```
Comment (jelilat) [inline jsonrpc/server.go:123]:
```suggestion
	if handlerT.Out(1) != reflect.TypeOf(&Error{}) {
```

**Review by jelilat — CHANGES_REQUESTED** [review]

File: jsonrpc/server.go (hunk 10 above)
Comment (omerfirmak) [reply jsonrpc/server.go]: done

File: jsonrpc/server.go (hunk 11 above)
Comment (omerfirmak) [reply jsonrpc/server.go]: done

File: rpc/types.go [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+	Status BlockStatus `json:"status"`
+	BlockHeader
+	Transactions []*Transaction `json:"transactions"`
+}
+
+// https://github.com/starkware-libs/starknet-specs/blob/a789ccc3432c57777beceaa53a34a7ae2f25fda0/api/starknet_api_openrpc.json#L1109
+type BlockWithTxHashes struct {
+	Status BlockStatus `json:"status"`
+	BlockHeader
+	TxnHashes []*felt.Felt `json:"transactions"`
```
Comment (jelilat) [inline rpc/types.go:115]: I think this should be `json:transaction_hashes`

Comment (omerfirmak) [reply rpc/types.go:115]: Nope, that is how spec names that field.

File: rpc/handlers_test.go [hunk 13]
```diff
@@ -0,0 +1,299 @@
...
+		assert.Equal(t, gwBlock.SequencerAddress, latestRpc.SequencerAddress)
+		assert.Equal(t, gwBlock.Timestamp, latestRpc.Timestamp)
+		for i := 0; i < len(gwBlock.Transactions); i++ {
+			assert.Equal(t, gwBlock.Transactions[i].Hash(), latestRpc.TxnHashes[i])
+		}
+
+		byHash, err := handler.GetBlockWithTxHashes(&rpc.BlockId{Hash: latestRpc.Hash})
+		assert.Equal(t, latestRpc, byHash)
+		byNumber, err := handler.GetBlockWithTxHashes(&rpc.BlockId{Number: latestRpc.Number})
+		assert.Equal(t, latestRpc, byNumber)
```
Comment (jelilat) [inline rpc/handlers_test.go:93]: We should probably handle the errors here

File: rpc/types.go [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+	Hash   *felt.Felt `json:"block_hash"`
+}
+
+// https://github.com/starkware-libs/starknet-specs/blob/a789ccc3432c57777beceaa53a34a7ae2f25fda0/api/starknet_api_openrpc.json#L814
+type BlockId struct {
+	Pending bool
+	Latest  bool
+	Hash    *felt.Felt
+	Number  uint64
+}
```
Comment (jelilat) [inline rpc/types.go:22]: Was thinking we should add `json` tag here as well, but not sure if it's intentional

```

type BlockId struct {
	Pending bool       `json:"pending,omitempty"`
	Latest  bool       `json:"latest,omitempty"`
	Hash    *felt.Felt `json:"hash,omitempty"`
	Number  uint64     `json:"number,omitempty"`
}
```

But the tests seem to be working fine after adding it

Comment (omerfirmak) [reply rpc/types.go:22]: Yeap, no tags was intentional. BlockId has custom unmarshaler. Check out how blockid is defined in the spec from the link above it.

Comment (jelilat) [reply rpc/types.go:22]: Right. Got it.

**Review by jelilat — APPROVED** [review]

File: jsonrpc/server.go (hunk 4 above)
Comment (IronGauntlets) [reply jsonrpc/server.go:111]: I tried playing around with it as well and apparently, the types need to match exactly.

File: jsonrpc/server.go [hunk 15]
```diff
@@ -0,0 +1,283 @@
...
+}
+
+// RegisterMethod verifies and creates an endpoint that the server recognizes.
+//
+// - name is the method name
+// - handler is the function to be called when a request is received for the
+// associated method. It should have (any, *jsonrpc.Error) as its return type
+// - paramNames are the names of parameters in the order that they are expected
+// by the handler
+func (s *Server) RegisterMethod(name string, params []Parameter, handler any) error {
```
Comment (IronGauntlets) [inline jsonrpc/server.go:111]: Is there a reason why the declaration is not the following:
```suggestion
func (s *Server) RegisterMethod(m Method) error {
```

File: jsonrpc/server.go [hunk 16]
```diff
@@ -0,0 +1,283 @@
...
+		} else {
+			reader.UnreadByte()
+			return char == '['
+		}
+	}
+
+	return false
+}
+
+func isNil(i interface{}) bool {
```
Comment (IronGauntlets) [inline jsonrpc/server.go:198]:
```suggestion
func isNil(i any) bool {
```

File: jsonrpc/server.go (hunk 15 above)
Comment (omerfirmak) [reply jsonrpc/server.go:111]: Probably because `RegisterMethod` predates `Method`, will change

File: jsonrpc/server_pkg_test.go [hunk 17]
```diff
@@ -0,0 +1,312 @@
...
+	"bytes"
+	"io"
+	"reflect"
+	"testing"
+
+	"github.com/NethermindEth/juno/core/felt"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
```
Comment (IronGauntlets) [inline jsonrpc/server_pkg_test.go:14]: All of the following tests can be moved to `server_test.go`. These private functions can all be tested through the `HandleRequest()`. We should not be testing private functions unless we require a complicated setup. You are already defining handlers and incoming requests so I don't see why they cannot be moved to `server_test.go`.

Also, all these tests require names and should be run as sub-tests.

**Review by IronGauntlets — CHANGES_REQUESTED** [review]

File: jsonrpc/server_pkg_test.go (hunk 17 above)
Comment (omerfirmak) [reply jsonrpc/server_pkg_test.go:14]: Those were the tests I used while developing that functions, it will take some time to make sure we dont lose coverage on the edge cases while moving them.

Comment (IronGauntlets) [reply jsonrpc/server_pkg_test.go:14]: You can note the coverage percentage before moving and then make sure it is the same after moving them. Goland can easily help with that.

Comment (IronGauntlets) [reply jsonrpc/server_pkg_test.go:14]: I just ran it with jsonrpc package with  coverage and it sits at 94.4%

File: jsonrpc/server_pkg_test.go [hunk 18]
```diff
@@ -0,0 +1,312 @@
...
+	tests := []struct {
+		handler    any
+		req        string
+		shouldFail bool
+		errorMsg   string
+	}{
+		{
+			handler: func(param1 int, param2 *arg) (any, error) {
+				assert.Equal(t, 44, param1)
```
Comment (IronGauntlets) [inline jsonrpc/server_pkg_test.go:159]: Once you run these tests as subtests please make sure `t` refers to the correct test. Currently, `t` refers to the parent tests.

File: node/node.go [hunk 19]
```diff
@@ -97,6 +110,7 @@ func (n *Node) Run(ctx context.Context) (err error) {
 		<-ctx.Done()
 		n.log.Infow("Shutting down Juno...")
 	}()
+	n.http.Run(ctx)
 	return n.synchronizer.Run(ctx)
```
Comment (jelilat) [inline node/node.go:114]: I think this should help us manage the http errors

```suggestion
	syncErrChan := make(chan error)
	httpErrChan := make(chan error)

	go func() {
		syncErrChan <- n.synchronizer.Run(ctx)
	}()
	go func() {
		httpErrChan <- n.http.Run(ctx)
	}()

	syncErr := <-syncErrChan
	httpErr := <-httpErrChan

	if syncErr != nil {
		return syncErr
	}
	if httpErr != nil {
		return httpErr
	}
	return err
```

Comment (omerfirmak) [reply node/node.go:114]: We dont want Run to block tho

Comment (omerfirmak) [reply node/node.go:114]: Run should only return an error if the service fails to start, so there is no reason to push them thru a channel because it will return immediately.

Any errors that the service may encounter in runtime should be handled within that service.

Comment (omerfirmak) [reply node/node.go:114]: There is not much the Node can do if http or synchronizer fails anyways

File: rpc/types.go [hunk 20]
```diff
@@ -0,0 +1,116 @@
+package rpc
+
+import (
+	"encoding/json"
+	"errors"
+
+	"github.com/NethermindEth/juno/core/felt"
+)
+
```
Comment (IronGauntlets) [inline rpc/types.go:9]: Please separate these types into their own files so it is easier to quickly identify which file contains what.

File: rpc/types_test.go [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+			},
+		},
+		"hash": {
+			blockIdJson: `{ "block_hash" : "0x123" }`,
+			expectedBlockId: rpc.BlockId{
+				Hash: new(felt.Felt).SetUint64(0x123),
+			},
+		},
+	}
+	t.Parallel()
```
Comment (IronGauntlets) [inline rpc/types_test.go:42]: I believe this usage is incorrect you should move it to line 45. unless this was intentional.

File: rpc/handlers_test.go [hunk 22]
```diff
@@ -0,0 +1,299 @@
...
+	canceler()
+	<-syncNodeChan
+}
+
+// implements only GetTransactionByHash from BlockchainReader interface
+// calling any other function from BlockchainReader will panic
+type txGwReader struct {
+	*blockchain.Blockchain
+	gw starknetdata.StarknetData
+}
```
Comment (IronGauntlets) [inline rpc/handlers_test.go:121]: Please rename this to fake blockchain reader

File: jsonrpc/server.go [hunk 23]
```diff
@@ -0,0 +1,279 @@
...
+			}
+		} else if !configuredParams[idx].Optional {
+			return nil, errors.New("missing non-optional param")
+		}
+
+		args = append(args, valueContainer.Elem())
+	}
+
+	return args, nil
+}
```
Comment (IronGauntlets) [inline jsonrpc/server.go:279]: Some comments would be helpful. It took me a while to understand what was happening here.

File: node/node.go [hunk 24]
```diff
@@ -76,12 +80,21 @@ func New(cfg *Config) (StarknetNode, error) {
...
 		blockchain:   chain,
 		synchronizer: synchronizer,
+		http: jsonrpc.NewHttp(cfg.RpcPort, []jsonrpc.Method{
+			{"starknet_chainId", nil, rpcHandler.ChainId},
+			{"starknet_blockNumber", nil, rpcHandler.BlockNumber},
+			{"starknet_blockHashAndNumber", nil, rpcHandler.BlockNumberAndHash},
+			{"starknet_getBlockWithTxHashes", []jsonrpc.Parameter{{Name: "block_id"}}, rpcHandler.GetBlockWithTxHashes},
+			{"starknet_getBlockWithTxs", []jsonrpc.Parameter{{Name: "block_id"}}, rpcHandler.GetBlockWithTxs},
+			{"starknet_getTransactionByHash", []jsonrpc.Parameter{{Name: "transaction_hash"}}, rpcHandler.GetTransactionByHash},
+		}),
```
Comment (IronGauntlets) [inline node/node.go:97]: registration of methods should be extracted into a function

**Review by IronGauntlets — CHANGES_REQUESTED** [review]

File: jsonrpc/server_test.go [hunk 25]
```diff
@@ -0,0 +1,153 @@
...
+		t.Run(desc, func(t *testing.T) {
+			err := server.RegisterMethod(jsonrpc.Method{"method", test.paramNames, test.handler})
+			assert.EqualError(t, err, test.want, desc)
+		})
+	}
+
+	t.Run("should not fail", func(t *testing.T) {
+		err := server.RegisterMethod(jsonrpc.Method{"method", nil, func(param1, param2 int) (int, *jsonrpc.Error) { return 0, nil }})
+		assert.NoError(t, err)
+	})
```
[the remark this answers is not in the document]
Comment (omerfirmak) [reply jsonrpc/server_test.go:56]: Probably not intentional but is valid. If the server only means to use list parameters it can omit parameter names.

File: jsonrpc/server.go [hunk 26]
```diff
@@ -0,0 +1,279 @@
...
+		if paramType.Kind() != reflect.Slice && paramType.Kind() != reflect.Map {
+			return errors.New("params should be an array or an object")
+		}
+	}
+
+	if r.Id != nil {
+		idType := reflect.TypeOf(r.Id)
+		floating := idType.Name() == "Number" && strings.Contains(r.Id.(json.Number).String(), ".")
+		if (idType.Kind() != reflect.String && idType.Name() != "Number") || floating {
+			return errors.New("id should be a string or an integer")
```
[the remark this answers is not in the document]
Comment (omerfirmak) [reply jsonrpc/server.go:75]: we allow `nil` id, see L71

File: jsonrpc/server.go [hunk 27]
```diff
@@ -0,0 +1,279 @@
...
+	Version string `json:"jsonrpc"`
+	Result  any    `json:"result,omitempty"`
+	Error   *Error `json:"error,omitempty"`
+	Id      any    `json:"id"`
+}
+
+type Error struct {
+	Code    int    `json:"code"`
+	Message string `json:"message"`
+	Data    any    `json:"data,omitempty"`
```
[the remark this answers is not in the document]
Comment (omerfirmak) [reply jsonrpc/server.go:41]: This is a generic JSONRPC package, so starknet related comments shouldnt go here.

`Data` is produced by handlers and server just passes them along in the response. There is not much it can do with it anyways.

File: jsonrpc/server.go [hunk 28]
```diff
@@ -0,0 +1,279 @@
...
+			if len(paramsList) != paramCount {
+				return nil, errors.New("missing param in list")
+			}
+
+			requestValue = paramsList[idx]
+		case reflect.Map:
+			paramsMap := params.(map[string]any)
+			if len(configuredParams) != paramCount {
+				return nil, errors.New("missing param name")
+			}
```
[the remark this answers is not in the document]
Comment (omerfirmak) [reply jsonrpc/server.go:255]: > Why would we allow registration of a method whose handler type input length differs from the length of parameters the caller provided?

yeah, I dont think there is a reason we should, removed

File: jsonrpc/server_test.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (omerfirmak) [reply jsonrpc/server_test.go:56]: ^ removed this feature

File: jsonrpc/server.go [hunk 29]
```diff
@@ -0,0 +1,279 @@
...
+	} else {
+		batchReq := batch{}
+		batchRes := []json.RawMessage{}
+
+		if err := json.NewDecoder(bufferedReader).Decode(&batchReq); err != nil {
+			return nil, err
+		}
+
+		for _, rawReq := range batchReq {
+			if res, err := s.Handle(rawReq); err == nil { // todo: handle async
```
[the remark this answers is not in the document]
Comment (omerfirmak) [reply jsonrpc/server.go:166]: great catch, thanks

File: rpc/types.go (hunk 20 above)
Comment (omerfirmak) [reply rpc/types.go:9]: What should the separation be based on? Different endpoints might share same data types.

File: jsonrpc/server_pkg_test.go (hunk 17 above)
Comment (omerfirmak) [reply jsonrpc/server_pkg_test.go:14]: done

File: jsonrpc/server.go (hunk 29 above)
[the remark this answers is not in the document]
Comment (omerfirmak) [reply jsonrpc/server.go:166]: ended up being a big refactor

**Review by IronGauntlets — APPROVED** [review]
Please merge #594 before closing this pr

**Comment by omerfirmak** [comment]
I cherry-picked and squashed first and last commit of #594
I really dont like using a wait group and explicitly handling each service we start, I propose we leave it to another PR so that we can discuss it further.

**Comment by IronGauntlets** [comment]
>I am reluctant to merge the second one, I propose we leave it to another PR so that we can discuss it further.

There was a possibility of go-routine leaks in jsonrpc.http.Run() functions and the documentation of Serve() mentioned that the application should not exit without Shutdown() returning. We relied on the fact that synchroniser will take longer to exit that jsonrpc.http.Run()

**Comment by IronGauntlets** [comment]
Happy for you to merge btw

**Comment by omerfirmak** [comment]
> > I am reluctant to merge the second one, I propose we leave it to another PR so that we can discuss it further.
> 
> There was a possibility of go-routine leaks in jsonrpc.http.Run() functions and the documentation of Serve() mentioned that the application should not exit without Shutdown() returning. We relied on the fact that synchroniser will take longer to exit than jsonrpc.http.Run().

Yeah, we should add a way for Node to handle lifecycles of the services it uses. But I cant seem to find the doc that mentions that, could you share it with me?

**Comment by IronGauntlets** [comment]
>Yeah, we should add a way for Node to handle lifecycles of the services it uses. But I cant seem to find the doc that mentions that, could you share it with me?

https://cs.opensource.google/go/go/+/refs/tags/go1.20.1:src/net/http/server.go;drc=ea4631cc0cf301c824bd665a7980c13289ab5c9d;l=2760-2762
